File: nexus/src/main/kotlin/borg/dgm/polyglot/TestSpec.kt
package borg.dgm.polyglot

import kotlinx.serialization.Serializable
import java.io.File

@Serializable
data class TestSpec(
    val instanceId: String,
    val repo: String,
    val repoScriptList: List<String>,
    val evalScriptList: List<String>,
    val envScriptList: List<String>,
    val arch: String
) {
    val setupEnvScript: String
        get() = """
            #!/bin/bash
            set -euxo pipefail
            ${envScriptList.joinToString("\n")}
        """.trimIndent()

    val evalScript: String
        get() = """
            #!/bin/bash
            set -uxo pipefail
            ${evalScriptList.joinToString("\n")}
        """.trimIndent()

    val installRepoScript: String
        get() = """
            #!/bin/bash
            set -euxo pipefail
            ${repoScriptList.joinToString("\n")}
        """.trimIndent()

    val baseImageKey: String
        get() = "pb.base.$arch:latest"

    val envImageKey: String
        get() = "pb.env.$instanceId:latest"

    val instanceImageKey: String
        get() = "pb.instance.$instanceId:latest"

    fun getInstanceContainerName(runId: String): String =
        "pb-container-${instanceId.replace("__", "-")}-$runId"

    companion object {
        fun fromEntry(entry: Map<String, Any>): TestSpec {
            val instanceId = entry["instance_id"] as String
            val repo = entry["repo"] as String
            val language = entry["language"] as String

            // Get language-specific test commands
            val testCommands = when (language) {
                "python" -> listOf("pytest -rA --tb=long")
                "rust" -> listOf("cargo test -- --include-ignored")
                "go" -> listOf("go test ./...")
                "javascript" -> listOf(
                    "set -e",
                    "[ ! -e node_modules ] && ln -s /npm-install/node_modules .",
                    "[ ! -e package-lock.json ] && ln -s /npm-install/package-lock.json .",
                    "sed -i 's/\\bxtest(/test(/g' *.spec.js",
                    "npm run test"
                )
                "cpp" -> listOf(
                    "set -e",
                    "[ ! -d \"build\" ] && mkdir build",
                    "cd build",
                    "cmake -DEXERCISM_RUN_ALL_TESTS=1 -G \"Unix Makefiles\" ..",
                    "make",
                    "cd ../"
                )
                "java" -> listOf("./gradlew test")
                else -> throw IllegalArgumentException("Unsupported language: $language")
            }

            // Get language-specific environment setup
            val envSetup = when (language) {
                "python" -> listOf(
                    "python -m pip install --upgrade pip",
                    "pip install -r requirements.txt"
                )
                "rust" -> listOf(
                    "rustup update",
                    "cargo build"
                )
                "go" -> listOf(
                    "go mod download",
                    "go build ./..."
                )
                "javascript" -> listOf(
                    "npm install"
                )
                "cpp" -> listOf(
                    "apt-get update",
                    "apt-get install -y build-essential cmake"
                )
                "java" -> listOf(
                    "./gradlew build"
                )
                else -> throw IllegalArgumentException("Unsupported language: $language")
            }

            // Get repository setup commands
            val repoSetup = listOf(
                "git clone $repo .",
                "git checkout ${entry["commit"]}"
            )

            return TestSpec(
                instanceId = instanceId,
                repo = repo,
                repoScriptList = repoSetup,
                evalScriptList = testCommands,
                envScriptList = envSetup,
                arch = "linux/amd64" // Default to amd64, can be overridden
            )
        }
    }
} 

File: dgm/swe_bench/test_spec.py
import os
from dataclasses import dataclass
from typing import List, Optional

@dataclass
class TestSpec:
    instance_id: str
    repo: str
    repo_script_list: List[str]
    eval_script_list: List[str]
    env_script_list: List[str]
    arch: str = "x86_64"

    def get_instance_container_name(self, run_id: str) -> str:
        return f"dgm_{self.instance_id}_{run_id}"

    def get_env_script(self) -> str:
        return "\n".join(self.env_script_list)

    def get_eval_script(self) -> str:
        return "\n".join(self.eval_script_list)

    def get_repo_script(self) -> str:
        return "\n".join(self.repo_script_list)

def make_test_spec(entry: dict) -> TestSpec:
    """
    Create a TestSpec object from a dataset entry.
    
    Args:
        entry: Dictionary containing test specification data
        
    Returns:
        TestSpec object with the test specification
    """
    instance_id = entry["instance_id"]
    repo = entry["repo"]
    
    # Get the language-specific test commands
    language = entry.get("language", "python").lower()
    if language == "python":
        test_cmd = "python -m pytest"
    elif language == "rust":
        test_cmd = "cargo test"
    elif language == "go":
        test_cmd = "go test ./..."
    elif language == "javascript":
        test_cmd = "npm test"
    elif language == "c++":
        test_cmd = "make test"
    elif language == "java":
        test_cmd = "mvn test"
    else:
        test_cmd = "python -m pytest"  # Default to Python
    
    # Create the evaluation script
    eval_script = f"""#!/bin/bash
set -e
cd /testbed
{test_cmd}
"""
    
    # Create the environment setup script
    env_script = f"""#!/bin/bash
set -e
cd /testbed
"""
    
    # Create the repository setup script
    repo_script = f"""#!/bin/bash
set -e
cd /testbed
git clone {repo} .
"""
    
    return TestSpec(
        instance_id=instance_id,
        repo=repo,
        repo_script_list=[repo_script],
        eval_script_list=[eval_script],
        env_script_list=[env_script],
    ) 